vql/linux/cronsnoop/cron-lib.go

```diff
@@ -0,0 +1,395 @@
+	if line_fields[0][0] == '#' || len(line_fields) == 0 {
+		return true
+	}
+
+	// if line starts with @ special string instead of ordinary cron time/date
+	// fields don't skip it.
+	if special_cron_string(line_fields) {
+		return false
+	} else if len(line_fields) <= 5 {
+		return true
+	}
+
+	return false
+}
+
+func (snooper *CronSnooper) parse_user_cron_file(path string) error {
+
+	var existing_user_cmds, current_commands []string
+	var cmd string
+	var ok bool
+
+	user := filepath.Base(path)
+
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+
+	// initialize user entries on first user modification detectioon
+	if existing_user_cmds, ok = snooper.user_cron_registry[user]; !ok {
+		snooper.user_cron_registry[user] = existing_user_cmds
+	}
+
+	for scanner.Scan() {
+		line := strings.TrimLeft(scanner.Text(), " \t")
+		split_line := strings.Split(line, " ")
+
+		if should_skip_line(split_line) {
+			continue
+		}
+
+		// Handle @-based entries in cron file
+		if special_cron_string(split_line) {
+			cmd = strings.Join(split_line[1:], " ")
+		} else {
+			cmd = strings.Join(split_line[5:], " ")
+		}
+
+		current_commands = append(current_commands, cmd)
+		// only take note of unique commands
+		if !exists(existing_user_cmds, cmd) {
+			existing_user_cmds = append(existing_user_cmds, cmd)
+			snooper.emit_event(cmd, user, path, Added)
+		}
+	}
+
+	// Filter out deleted entries
+	existing, deleted := filter_deleted_cmds(existing_user_cmds, current_commands)
+
+	// emit events
+	for _, deleted_cmd := range deleted {
+		snooper.emit_event(deleted_cmd, user, path, Deleted)
+	}
+
+	// user got all of its entries deleted so just delete the user altogether
+	if len(existing) == 0 {
+		//TODO: Emit an event that all of user's entries are deleted
+		delete(snooper.user_cron_registry, user)
+	} else {
+		// We have some entries left for the user and filtered contains the final
+		// "view" of the user's cron entries
+		snooper.user_cron_registry[user] = existing
+	}
+
+	if err := scanner.Err(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (snooper *CronSnooper) parse_system_cron_file(path string) error {
+
+	var cmd, user string
+	var cron_registry map[string][]string
+	var ok bool
+
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+
+	parsed_commands := make(map[string][]string)
+
+	// initialize hashmap for the given file
+	if cron_registry, ok = snooper.system_cron_registry[path]; !ok {
+		cron_registry = make(map[string][]string)
+		snooper.system_cron_registry[path] = cron_registry
+	}
+
+	for scanner.Scan() {
+		line := strings.TrimLeft(scanner.Text(), " \t")
+		split_line := strings.Split(line, " ")
+
+		if should_skip_line(split_line) {
+			continue
+		}
+
+		// TODO: Probably could optimise how we get the command instead of doing
+		// split + join
+		if special_cron_string(split_line) {
```

Comment:
Maybe, but this isn't really performance-critical code, is it?

Reply:
On a second pass you are right.